docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,802 @@
+other lifecycle moments.
+
+This proposal aims to solve the above problem in a more structured and generic way, by introducing the Runtime SDK,
+a set of rules, recommendations and fundamental capabilities required to implement a new extensibility mechanism
+that will allow systems, products and services built on top of Cluster API to interact with the Workload Cluster’s
+lifecycle. 
+
+The key elements of the above extensibility mechanism are Runtime Hooks and Runtime Extensions.
+
+Runtime Hooks and Runtime Extensions are powerful and flexible, and _by opportunity_ it will be also possible to
+use this capability for allowing the user to hook into Cluster API reconcile loops at "low level", e.g. 
+by allowing to plugin external patches to be executed on every topology reconcile.
+
+### Goals
+
+To define the Runtime SDK and more specifically
+
+- To define the rules ensuring Runtime Hooks can evolve over time:
+  - When/how to modify the current version;
+  - When/how to create a new version;
+  - When/how to deprecate an old version, as well as mechanisms to inform users about versions being deprecated;
+  - When/how to drop an old version, as well as providing a mechanism to prevent users to upgrade Cluster API when
+    this will break installed Runtime Extensions;
+- To define the fundamental capabilities/tooling to be implemented in CAPI in order to allow the implementation of
+  Runtime Hooks.
+- To provide an initial set of guidelines for Runtime Extension developers.
+- To define how external Runtime Extensions can be registered within the Cluster API Runtime.
+
+### Non-Goals
+
+- To identify or specify the list of Runtime Hooks that should be implemented; some examples of possible Runtime Hooks
+  will be eventually provided, but it is not in the scope of this document to define them in detail;
+- To replace controllers or any other component of the Cluster API Runtime (including infrastructure providers,
+  bootstrap providers, control plane providers and the CRD/contract based extension mechanism they rely on).
+- To support the implementation of Runtime Hooks anywhere else outside the main Cluster API repository.
+
+### Future Work
+
+- Identify and specify the list of Runtime Hooks to be implemented; this will be addressed iteratively by a set of
+  future proposals, all of them building on top of the foundational capabilities introduced by this document;
+- Eventually consider deprecation of machine deletion hooks and replacement with a Runtime Hook;
+- Improve the Runtime Extension developer guide based on experience and feedback;
+- Add metrics about Runtime Extension calls (usage, usage vs deprecated versions, duration, error rate etc.);
+- Allow providers to use the same SDK to define their own hooks.
+
+## Proposal
+
+### User Stories
+
+- As a cluster operator I want to be able to execute a particular action in well-defined moments of the Workload
+  Cluster’s lifecycle, e.g. 
```

Comment:
I have considered breaking the last point into a separate user story but then dropped the idea because reconciliation at the end is an implementation detail from the user PoV "computing the Cluster topology" could be considered a Cluster Lifecycle event that happens continuously.